test(client): add `relationMode` datasource property tests for all databases

Jolg42 opened the pull request:
Related: https://github.com/prisma/prisma/issues/10806

```sh
RELATION_MODE=foreignKeys pnpm run test:functional:code --relation-mode-tests-only
RELATION_MODE=prisma pnpm run test:functional:code --relation-mode-tests-only
pnpm run test:functional:code --relation-mode-tests-only

# Or only one relation type by providing the file name:
pnpm run test:functional:code --relation-mode-tests-only relationMode/tests_m-to-n.ts
```

<details>
<summary>How to run - Previously (early commit)</summary>

```sh
RELATION_MODE=foreignKeys pnpm run test:functional:code relationMode
RELATION_MODE=prisma pnpm run test:functional:code relationMode
pnpm run test:functional:code relationMode

# Or only one relation type by providing the file name:
pnpm run test:functional:code relationMode/tests_m-to-n.ts
```

and before

```sh
RI=foreignKeys pnpm run test:functional:code referential-integrity
RI=prisma pnpm run test:functional:code referential-integrity
pnpm run test:functional:code referential-integrity

# Or only one relation type by providing the file name:
pnpm run test:functional:code referential-integrity/tests_m-to-n.ts
```

</details>

Internal Notion pages
- [Project](https://www.notion.so/prismaio/Referential-Actions-TS-Test-Plan-52571af2aa6a48c1ae3a929880fd25aa)
- [Findings](https://www.notion.so/prismaio/Phase-1-Report-on-findings-f21c7bb079c5414296286973fdcd62c2)

Jolg42 committed:
test(client): prototype referential integrity tests

Related: https://github.com/prisma/prisma/issues/10806

Test Suites: 1 failed, 1 total
Tests:       3 failed, 45 passed, 48 total
Snapshots:   0 total
Time:        9.838 s

Jolg42 committed:
exclude mongodb to see if types tests fail like locally

Jolg42 commented:
@tomhoule maybe something you might know about.

This is the tests failing:

postgresql
```
  {
    name: 'findMany on column_name_that_becomes_empty_string',
    up: `
        CREATE TABLE column_name_that_becomes_empty_string (
          field1 serial primary key not null,
          "12345" int DEFAULT NULL
        );

        create type invalid_enum as enum ('Y','N','123','$§!');

        CREATE TABLE invalid_enum_value_name (
          field1 serial primary key not null,
          here_be_enum invalid_enum DEFAULT NULL
        );

        CREATE TABLE no_unique_identifier (
          field1 int DEFAULT NULL,
          field2 int DEFAULT NULL
        );

        CREATE TABLE unsupported_type (
          field1 serial primary key not null,
          unsupported polygon DEFAULT NULL
        );
      `,
    do: async (client) => {
      return await client.column_name_that_becomes_empty_string.findMany({})
    },
    expect: [],
  },
```

mariadb
```
  {
    name: 'findMany on column_name_that_becomes_empty_string',
    up: `
        CREATE TABLE \`column_name_that_becomes_empty_string\` (
          \`field1\` int(11) NOT NULL AUTO_INCREMENT,
          \`12345\` int(11) DEFAULT NULL,
          PRIMARY KEY (\`field1\`)
        ) ENGINE=InnoDB DEFAULT CHARSET=latin1;

        CREATE TABLE \`invalid_enum_value_name\` (
          \`field1\` int(11) NOT NULL AUTO_INCREMENT,
          \`here_be_enum\` enum('Y','N','123','$§!') DEFAULT NULL,
          PRIMARY KEY (\`field1\`)
        ) ENGINE=InnoDB DEFAULT CHARSET=latin1;

        CREATE TABLE \`no_unique_identifier\` (
          \`field1\` int(11) DEFAULT NULL,
          \`field2\` int(11) DEFAULT NULL
        ) ENGINE=InnoDB DEFAULT CHARSET=latin1;

        CREATE TABLE \`unsupported_type\` (
          \`field1\` int(11) NOT NULL AUTO_INCREMENT,
          \`unsupported\` binary(50) DEFAULT NULL,
          PRIMARY KEY (\`field1\`)
        ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
      `,

    do: async (client) => {
      return await client.column_name_that_becomes_empty_string.findMany({})
    },
    expect: [],
  },
```

Jolg42 committed:
wip rewrite with Alberto

jkomyno commented:
Note: `referentialIntegrity: "prisma"` already shows differences w.r.t. `referentialIntegrity: "foreignKeys"`:
- `1:1 mandatory (explicit) › [create] child with non existing parent should throw` doesn't throw a foreign key constraint violation error
- slightly different logs (e.g., mentioning the `userId` column rather than `id`)

jkomyno committed:
chore: update tests

Jolg42 commented on packages/client/tests/functional/referential-integrity/tests.ts:
```diff
@@ -343,16 +384,263 @@ testMatrix.setupTestSuite(
+          },
+        })
+
+        // TODO: is this a bug? Probably it's trying to disconnect under the hood?
+        try {
+          await prisma[userModel].update({
+            where: { id: '1' },
+            data: {
+              id: '3',
+              profile: {
+                connect: { id: '1' },
+              },
+            },
+          })
+          assertUnreachable()
+        } catch (error) {
+          expect(error.message).toContain('The change you are trying to make would violate the required relation \'ProfileOneToOneToUserOneToOne\' between the `ProfileOneToOne` and `UserOneToOne` models.')
+        }
+      })
```
I have no idea about what you mean here. Are you saying it behaves differently with `referentialIntegrity = "prisma"` vs `referentialIntegrity = "foreignKeys"`, if so what's the difference?

Jolg42 replied:
Note: no need to investigate now, we need to finish a first version of the test now

Jolg42 commented on packages/client/tests/functional/referential-integrity/tests.ts:
```diff
@@ -0,0 +1,677 @@
+          userId: '1',
+        })
+
+        const user1WithUpdatedProfile = await prisma[userModel].findUniqueOrThrow({
+          where: { id: '1' },
+          include: { profile: true },
+        })
+        expect(user1WithUpdatedProfile).toEqual({
+          id: '1',
+          profile: {
+            id: '3',
+            userId: '1',
+          },
+        })
+
+        const profiles = await prisma[profileModel].findMany({
+          orderBy: [{ id: 'asc' }],
+        })
+        expect(profiles).toEqual([
+          {
+            id: '2',
+            userId: '2',
+          },
+          {
+            id: '3',
+            userId: '1',
+          },
+        ])
```
Using `findMany()` for each model, would simplify assertions in all tests, and show/verify what we expect to see for all the items (findUniqueOrThrow only verifies 1 item)

Jolg42 commented on packages/client/tests/functional/referential-integrity/tests.ts:
```diff
@@ -0,0 +1,677 @@
+          },
+        })
+
+        try {
+          await prisma[userModel].updateMany({
+            data: { id: '2' }, // existing id
+            where: { id: '1' },
+          })
+          assertUnreachable()
+        } catch (error) {
+          expect(error.message).toContain('Unique constraint failed on the fields: (\`id\`)')
+        }
```
Can be simplified to
```suggestion
        await expect(
          prisma[userModel].updateMany({
            where: { id: '1' },
            data: {
              id: '2', // existing id
            },
          }),
        ).rejects.toThrowError('Unique constraint failed on the fields: (`id`)')
```

Jolg42 reviewed (commented):
Note: we need to `await` `expect(...).rejects/resolves.` (see linting errors)

Jolg42 commented on packages/client/tests/functional/referential-integrity/tests.ts:
```diff
@@ -0,0 +1,677 @@
+      })
+
+      test('[nested] update parent id [update] child should succeed', async () => {
+        await prisma[userModel].create({
+          data: {
+            id: '1',
+            profile: {
+              create: { id: '1' },
+            },
+          },
+        })
+        await prisma[userModel].create({
+          data: {
+            id: '2',
+            profile: {
+              create: { id: '2' },
+            },
+          },
+        })
```
We should remove these copied and pasted lines everywhere and use a small function that does the same.
It would be easier to see what we actually want to test in these tests, and remove potential copied and pasted typos.

Jolg42 commented on packages/client/tests/functional/referential-integrity/tests.ts:
```diff
@@ -0,0 +1,677 @@
+          },
+        })
+
+        try {
+          await prisma[userModel].update({
+            where: { id: '1' },
+            data: {
+              id: '3',
+              profile: {
+                disconnect: true,
+              },
+            },
+          })
+          assertUnreachable()
+        } catch (error) {
+          expect(error.message).toContain('The change you are trying to make would violate the required relation \'ProfileOneToOneToUserOneToOne\' between the `ProfileOneToOne` and `UserOneToOne` models.')
+        }
```
Assertion can be rewritten, see previous comment

jkomyno committed:
chore: used .rejects.toThrowError

jkomyno committed:
chore: started adding ref actions

jkomyno committed:
chore: mongodb only supports 'referentialIntegrity: prisma'

jkomyno committed:
chore: added support for conditional error messages

Jolg42 committed:
wip

reorg with describe
revert matrix and schema changes
beforeEach check and create users with profiles

Jolg42 committed:
fix referentialIntegrityLine

Jolg42 committed:
add skeleton for 1:n and m:n tests

jkomyno committed:
chore: highlighted some test errors with mongodb

jkomyno committed:
chore: added create tests for 1:n relationship

jkomyno committed:
chore: added TODO comment

Jolg42 committed:
WIP m:n for SQL databases

Jolg42 committed:
ci: try running github actions matrix with our tests only

Jolg42 committed:
ci: remove needs: detect_jobs_to_run

Jolg42 committed:
ci: fix os matrix

Jolg42 committed:
ci: small CI env var tweak

Jolg42 committed:
ci: fix bash line

jkomyno committed:
chore: added some update/delete tests for 1:n

jkomyno committed:
chore: added comments to _matrix.ts

janpio commented on packages/client/tests/functional/referential-integrity/_matrix.ts:
```diff
@@ -9,6 +9,10 @@ if (RI && RI !== 'prisma' && RI !== 'foreignKeys') {
 type RIType = 'prisma' | 'foreignKeys' | ''
 const referentialIntegrity: RIType = (RI as RIType) || ''
 
+// Note: testing 'SetDefault' requires a relation with a scalar field having the "@default" attribute.
+// If no defaults are provided for any of the scalar fields, a runtime error will be thrown.
+const referentialActionsChoices = ['', 'Cascade', 'Restrict', 'NoAction', 'SetNull']
```
💡 I was not aware of that. Interesting.

Jolg42 committed:
postgresql m:n all referential actions

Jolg42 committed:
merge main

Jolg42 committed:
Merge remote-tracking branch 'origin/main' into joel/referential-integrity

Jolg42 committed:
DEFAULT, exclude Restrict from SQL Server, fix tests

Jolg42 committed:
add mongodb tests for m:n

jkomyno committed:
chore: added some additional ref actions, added support for SQL Server

jkomyno committed:
chore: merge upstream, fix conflicts

jkomyno committed:
chore: fixed bug with DEFAULT not tested, added upsert test, fixed some conditional errors, found panic on SQLServer with SetDefault

Jolg42 committed:
reuse sql schema for mongodb 1:1 1:n

Jolg42 committed:
1:1 test all referral actions for onDelete & clean

jkomyno committed:
chore: improved ref actions, fixed failing tests

jkomyno committed:
Merge branch 'joel/referential-integrity' of https://github.com/prisma/prisma into joel/referential-integrity

jkomyno committed:
chore: reproduced issue #14271

jkomyno committed:
chore: split relationships in separate files

Jolg42 committed:
split m:n mongodb tests into its own file

Jolg42 committed:
add `enabled Boolean?` on 1-to-n

It's a reproduction of https://github.com/prisma/prisma/issues/13766

Jolg42 committed:
m:n add comments where RI=prisma - Cascade resolves instead of failing in update and create

Jolg42 committed:
add `published   Boolean?` to MongoDB and update tests

Jolg42 committed:
1:1 add `enabled Boolean?` and MongoDB update tests

Jolg42 committed:
add `published   Boolean?` to m:n with update tests

jkomyno committed:
chore: updated sql server tests

jkomyno committed:
Merge branch 'joel/referential-integrity' of https://github.com/prisma/prisma into joel/referential-integrity

Jolg42 committed:
add comments to m:n

[skip ci]

Jolg42 committed:
add a mongodb test for immutable _id

Jolg42 committed:
full database matrix for CI

Jolg42 commented on packages/client/tests/functional/jest.config.js:
```diff
@@ -21,7 +21,7 @@ module.exports = () => {
     snapshotSerializers: ['@prisma/internals/src/utils/jestSnapshotSerializer'],
     setupFilesAfterEnv: ['./_utils/setupFilesAfterEnv.ts'],
-    testTimeout: 40000,
+    testTimeout: 10_000,
```
@sonatype-lift ignoreall

Jolg42 committed:
split schemas into simple files

Jolg42 committed:
m to n MongoDB: change tests to match current implementation expectations

jkomyno committed:
chore: add JSON support to getTestSuiteFullName

jkomyno committed:
chore: moved ref integrity utils to its own folder. Added builder pattern to conditional error. Added matrix generation. Extracted self-contained utilities out of _schema.ts file.

jkomyno committed:
chore: added sqlite support and referentialIntegrity=prisma snapshots to 1:1 and 1:n relations

jkomyno committed:
chore: pnpm-lock

jkomyno committed:
chore: merge upstream

Jolg42 committed:
clean and add sqlite on m to n

Jolg42 committed:
m to n add RI=Prisma messages as comments + small edits

add expectation even for `toThrowError()` cases

Jolg42 committed:
m:n: split 1 test for prisma/foreignKeys

jkomyno committed:
chore: removed temp folder that wasn't supposed to be committed

jkomyno committed:
chore: imported right 'checkIfEmpty' util for m:n relation

jkomyno committed:
chore: added prisma snapshots and errors as comments in m:n relation (no MongoDB yet)

jkomyno committed:
chore: added mongodb and comments to _matrix

jkomyno committed:
fix: moved _referential-integrity-utils to _utils/referential-integrity to support  ERR_PNPM_NO_SCRIPT  Missing script: test:functional-code command

Jolg42 committed:
update matrix for MongoDB: add SetNull

jkomyno committed:
chore: added comments to m:n

Jolg42 committed:
m to n: small edits for RI=prisma

Jolg42 committed:
test 1:1 / 1:n add expects for expected throws and fix some RI=prisma messages

Jolg42 committed:
chore: prettier

Jolg42 committed:
1:1 handle error messages depending on RI=prisma/foreignKeys

Jolg42 committed:
* simplify conditionals

Jolg42 committed:
m to n: change expectations to match current state

Jolg42 committed:
chore: comment skipped test

jkomyno committed:
sqlite: skip createMany in 1:n relations

jkomyno committed:
chore: added missing sqlite snapshots to 1:n relation

Jolg42 committed:
1:n add missing delete action variants

Jolg42 committed:
1:n simplify and handle expected RI=prisma tests that should succeed

jkomyno committed:
chore: added reproduction of issue #10000

jkomyno committed:
chore: updated reproduction of issue #10000

jkomyno committed:
chore: reproduce issue #12378

Jolg42 committed:
Merge remote-tracking branch 'origin' into joel/referential-integrity

# Conflicts:
#	.github/workflows/test.yml
#	packages/client/tests/functional/jest.config.js

Jolg42 committed:
1:1 separate onUpdate: Restrict, NoAction

Jolg42 committed:
1:1 fix create test from should throw to should succeed

Jolg42 committed:
Merge remote-tracking branch 'origin/main' into joel/referential-integrity

# Conflicts:
#	packages/client/tests/functional/_utils/getTestSuiteInfo.ts
#	pnpm-lock.yaml

Jolg42 committed:
ci: separate action test for each relation / file

Jolg42 committed:
ci: continue-on-error: true

Jolg42 committed:
fix: referential actions test params

Jolg42 committed:
fix ts checks

Jolg42 committed:
only run pnpm run test:functional:code filename

Jolg42 committed:
fix Cannot destructure property 'onDelete' of 'suiteConfig.referentialActions' as it is undefined.

jkomyno committed:
chore: added reproduction of issue #12557

jkomyno committed:
Merge branch 'joel/referential-integrity' of github.com:prisma/prisma into joel/referential-integrity

jkomyno committed:
chore: add assertions to nested child connect in 1:1

Jolg42 committed:
chore: clean unused code and better describe title

Jolg42 committed:
Merge branch 'joel/referential-integrity' of github.com:prisma/prisma into joel/referential-integrity

Jolg42 committed:
Merge remote-tracking branch 'origin' into joel/referential-integrity

# Conflicts:
#	.github/workflows/test.yml
#	packages/client/package.json
#	pnpm-lock.yaml

Jolg42 committed:
test: add test about NO ACTION behavior with DEFERRABLE constraint

jkomyno committed:
chore: merge main, fix conflicts

Jolg42 committed:
Merge remote-tracking branch 'origin' into joel/referential-integrity

# Conflicts:
#	.github/workflows/test.yml

Jolg42 committed:
manually bump engines

Jolg42 committed:
chore: rename referentialIntegrity datasource property to relationMode

Jolg42 committed:
Merge remote-tracking branch 'origin' into joel/referential-integrity

# Conflicts:
#	packages/client/package.json
#	packages/engines/package.json
#	packages/fetch-engine/package.json
#	packages/migrate/package.json
#	pnpm-lock.yaml

Jolg42 committed:
fix github actions

Jolg42 committed:
fix test about 14759 and RELATION_MODE env var

Jolg42 committed:
Merge branch 'main' into joel/referential-integrity

Jolg42 committed:
cleanup some reproduce-x test cases

Jolg42 committed:
add sqlite error snapshot

Jolg42 committed:
remove continue-on-error

Jolg42 committed:
remove outdated comments

Jolg42 committed:
Merge remote-tracking branch 'origin' into joel/referential-integrity

# Conflicts:
#	pnpm-lock.yaml

Jolg42 committed:
ci: remove last continue-on-error: true

Jolg42 committed:
test: add `SetNull` to the matrix only for relationMode=prisma

Jolg42 committed:
ci: add continue-on-error: true to run all tests

Jolg42 committed:
fix snapshot with lowerace table name and run SQL Server

Jolg42 committed:
exclude SQL Server / SetNull when relationMode !== prisma

Jolg42 committed:
fix computeMatrix for SQL Server Restrict emulation which is not implemented

Jolg42 committed:
change failing tests (issue 15683) to use .failing

Jolg42 committed:
fix SQL Server test

Jolg42 committed:
test(reproductions): fix empty test name

Jolg42 committed:
fix relationMode = prisma SQL Server error snapshot

Jolg42 committed:
fix relationMode = prisma SQL Server error snapshot

Jolg42 committed:
fix relationMode = prisma SQL Server error snapshot

Jolg42 committed:
Merge branch 'main' into joel/referential-integrity

Jolg42 committed:
comment vitess docker image (unused) and put back timeout to 60s

Jolg42 committed:
test: fix snapshots for field-reference

Jolg42 committed:
test: fix snapshot and getTestSuiteParametersString

Jolg42 committed:
ignore expected type errors

Jolg42 committed:
add note about SetNull is not run for foreignKeys in the test suite

Jolg42 committed:
test: tweak comment and fix 1:1 test snapshots

Jolg42 committed:
Unused '@ts-expect-error' directive.

Jolg42 committed:
add `--relation-mode-tests-only` to unblock merge

Jolg42 committed:
Merge branch 'main' into joel/referential-integrity

janpio commented on packages/client/tests/functional/relationMode/_matrix.ts:
```diff
@@ -0,0 +1,64 @@
+import { computeMatrix } from '../_utils/relationMode/computeMatrix'
+
+const RelationModeEnv = process.env.RELATION_MODE
+if (RelationModeEnv && RelationModeEnv !== 'prisma' && RelationModeEnv !== 'foreignKeys') {
```
Shouldn't this also be able to be `''`?

Jolg42 replied:
It is, we just check if it's defined it must be either `prisma` or `foreignKeys`

janpio commented on packages/client/tests/functional/_utils/relationMode/computeMatrix.ts:
```diff
@@ -0,0 +1,69 @@
+    Providers.MYSQL,
+    Providers.SQLITE,
+  ] as const
+  //  Note SetDefault is not implemented in the emulation
```
Still true?

Jolg42 replied:
Yes, it was never implemented and same today
https://github.com/prisma/prisma-engines/blob/810aaacead8837bb6f5b6a051d719353d2c360fb/psl/psl-core/src/datamodel_connector/relation_mode.rs#L18-L23

janpio commented on packages/client/tests/functional/_utils/relationMode/computeMatrix.ts:
```diff
@@ -0,0 +1,69 @@
+  const providers = providersBase.filter((provider) => !(providersDenyList || []).includes(provider))
+
+  const referentialActionsDenylistByProvider = {
+    // 'Restrict' is not available when using 'sqlserver' as a provider, and it triggers a schema parsing error arising from DMMF.
+    [Providers.SQLSERVER]: ['Restrict'],
```
Still correct?

Jolg42 replied:
Yes it comes from the fact that SQL Server doesn't have `Restrict` as a valid action, it's in our docs too

janpio commented on packages/client/tests/functional/_utils/relationMode/computeMatrix.ts:
```diff
@@ -0,0 +1,69 @@
+    const denyList = referentialActionsDenylistByProvider[entry.provider] || []
+    let referentialActions = referentialActionsBase.filter((action) => !denyList.includes(action))
+
+    // Since `SetNull` when using Foreign Keys fails with migration errors on MySQL, CockroachDB and SQL Server
+    // We skip these combinations in the matrix (= filtering them out)
+    if (
+      relationMode !== 'prisma' &&
+      [Providers.MYSQL, Providers.COCKROACHDB, Providers.SQLSERVER].includes(entry.provider)
+    ) {
+      // For 1:1/1:n tests
+      //
+      // CockroachDB errors with:
+      // 1:1 - ERROR: cannot add a SET NULL cascading action on column "cl8y8lubt0000634p91cigsy4.public.ProfileOneToOne.userId" which has a NOT NULL constraint
+      // 1:n - ERROR: cannot add a SET NULL cascading action on column "cl8ya9ab30004xl4p4x7h8wuw.public.PostOneToMany.authorId" which has a NOT NULL constraint
+      // m:n - ERROR: cannot add a SET NULL cascading action on column "cl8yby0am0004qg4p6utsdh1i.public.CategoriesOnPostsManyToMany.postId" which has a NOT NULL constraint
+
+      // MySQL errors with:
+      // 1:1 - Can't create table `cl8y7xlzb000epx4p3o5oh6nu`.`ProfileOneToOne` (errno: 150 "Foreign key constraint is incorrectly formed")
+      // 1:n - Can't create table `cl8ya7mfd0004rq4p0uga2d6z`.`PostOneToMany` (errno: 150 "Foreign key constraint is incorrectly formed")
+      // m:n - Can't create table `cl8ybu3ni0004eo4p9xeubq4u`.`CategoriesOnPostsManyToMany` (errno: 150 "Foreign key constraint is incorrectly formed")
+      //
+      // SQL Server errors with:
+      // 1:1 - Cannot create the foreign key "ProfileOneToOne_userId_fkey" with the SET NULL referential action, because one or more referencing columns are not nullable.
+      // 1:n - Cannot create the foreign key "PostOneToMany_authorId_fkey" with the SET NULL referential action, because one or more referencing columns are not nullable.
+      // m:n - Cannot create the foreign key "CategoriesOnPostsManyToMany_postId_fkey" with the SET NULL referential action, because one or more referencing columns are not nullable.
+      referentialActions = referentialActions.filter((action) => action !== 'SetNull')
+    }
```
Shouldn't we technically confirm here that such a schema does not validate and instead get a validation error when trying to migrate? Or should that be somewhere else?

Jolg42 replied:
We usually do the schema validation errors tests in the engines repo

jkomyno replied:
The schema validation error tests in `prisma/prisma-engines` will be added as a result of https://github.com/prisma/prisma/issues/14673.

janpio commented on packages/client/tests/functional/_utils/relationMode/computeSchemaHeader.ts:
```diff
@@ -0,0 +1,41 @@
+}
+
+export function computeSchemaHeader({ provider, previewFeatures, relationMode }: ComputeSchemaHeader): string {
+  const USE_PLANETSCALE = false
```
Preparation for later?

Jolg42 replied:
Manual boolean, we need something else later

janpio commented on packages/client/tests/functional/_utils/relationMode/conditionalError.ts:
```diff
@@ -0,0 +1,70 @@
+  relationMode: RelationMode
+}
+
+// O.AtLeast<Record<Providers, string>>
```
?

jkomyno replied:
```suggestion
```

janpio commented on packages/client/tests/functional/_utils/relationMode/conditionalError.ts:
```diff
@@ -0,0 +1,70 @@
+}
+
+interface ConditionalErrorSnapshot {
+  // snapshot(errors: O.AtLeast<Record<RelationMode, O.AtLeast<Record<Providers, string>>>>): string
```
?

jkomyno replied:
```suggestion
```

janpio commented on packages/client/tests/functional/_utils/relationMode/conditionalError.ts:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    if (errorBase === undefined) {
+      return `TODO: add error for relationMode=${relationMode}`
+    }
+
+    return errorBase[provider] || `TODO: add error for provider=${provider}`
```
?

Jolg42 replied:
If we do not have a snapshot of the error message in the test, we output `TODO: add error for relationMode=${relationMode}`

So it's obvious what needs to be done

janpio commented on packages/client/tests/functional/relationMode/README.md:
```diff
@@ -0,0 +1,5 @@
+## Testing relationMode
+
+Note: We are not testing SetNull with `foreignKeys` because it is invalid.
+SetNull with a non-optionnal relation errors when the migration DDL is applied for all databases
+(except for PostgreSQL where it fails at runtime)
```
Is this still valid?

Jolg42 replied:
At the moment, yes.
Will change for a schema validation error after https://github.com/prisma/prisma/issues/14673 is done.

janpio replied:
Add link to issue and make sure this is updated when that is tackled please.

janpio commented on packages/client/tests/functional/relationMode/_matrix.ts:
```diff
@@ -0,0 +1,64 @@
+const relationMode: RelationMode = (RelationModeEnv as RelationMode) || ''
+
+// Note: testing 'SetDefault' requires a relation with a scalar field having the "@default" attribute.
+// If no defaults are provided for any of the scalar fields, a runtime error will be thrown.
```
Are we testing that both database and emulated to confirm this?
(Should this maybe be a validation error instead?)

Jolg42 replied:
SetDefault is invalid for `relationMode="prisma"`

janpio commented on packages/client/tests/functional/relationMode/_matrix.ts:
```diff
@@ -0,0 +1,64 @@
+//
+// Note: 'SetNull' is only available on optional relations.
+//
+// Note: 'SetDefault' is making SQL Server crash badly
+//  const referentialActionsChoices = ['', 'Cascade', 'NoAction']
```
?

Jolg42 replied:
We do not test SetDefault at all, see https://www.notion.so/prismaio/Phase-1-Report-on-findings-f21c7bb079c5414296286973fdcd62c2#ac4d9f6a5d3842b5b6ff5b877e7e6782

Jolg42 replied:
Since `SetDefault` is not valid when `relationMode="prisma"` there is no interest in doing that right now.

janpio commented on packages/client/tests/functional/relationMode/_matrix.ts:
```diff
@@ -0,0 +1,64 @@
+  ],
+  [
+    {
+      previewFeatures: '"referentialIntegrity"',
```
Still correct or should be `relationMode` now?

Jolg42 replied:
We can change it

Jolg42 replied:
Still correct, the preview feature flag didn't change, and won't

janpio commented on packages/client/tests/functional/relationMode/tests_1-to-1.ts:
```diff
@@ -0,0 +1,1413 @@
+          ).toEqual([])
+        })
+
+        test('[create] nested child [create]', async () => {
```
Is this title missing the expectation of the test? (Seems it should universally succeed)

janpio commented on packages/client/tests/functional/relationMode/tests_1-to-1.ts:
```diff
@@ -0,0 +1,1413 @@
+          ).toEqual([])
+        })
+
+        test('[create] nested child [create]', async () => {
+          await prisma[userModel].create({
+            data: {
+              id: '1',
+              profile: {
+                create: { id: '1' },
+              },
+            },
+          })
+
+          expect(
+            await prisma[userModel].findUniqueOrThrow({
+              where: { id: '1' },
+            }),
+          ).toEqual({
+            id: '1',
+            enabled: null,
+          })
+          expect(
+            await prisma[profileModel].findUniqueOrThrow({
+              where: { userId: '1' },
+            }),
+          ).toEqual({
+            id: '1',
+            userId: '1',
+            enabled: null,
+          })
+        })
```
Is this title missing the expectation of the test? (Seems it should universally succeed)

Jolg42 replied:
Fixed

Jolg42 committed:
Merge remote-tracking branch 'origin' into joel/referential-integrity

# Conflicts:
#	pnpm-lock.yaml

Jolg42 committed:
test: skip NoAction for PostgreSQL and SQLite

Related https://github.com/prisma/prisma-engines/pull/3274

Jolg42 committed:
fix relationMode value

Jolg42 committed:
fix test name + comments

Jolg42 committed:
cleanup matrix and move reproduce tests to issues directory

Jolg42 committed:
add test for referentialIntegrity="prisma"

Closes https://github.com/prisma/prisma/issues/15736

Jolg42 committed:
cleanup comments and misleading `toThrowError()` for expected to fail tests

jkomyno committed:
chore: removed useless comment

jkomyno committed:
chore: removed useless comments

Jolg42 committed:
fix imports for moved tests in issues directory

janpio commented on packages/client/tests/functional/relationMode/tests_m-to-n-MongoDB.ts:
```diff
@@ -0,0 +1,339 @@
+    /**
+     * m:n relationship
+     */
+    describeIf(isMongoDB)('m:n mandatory (explicit) - MongoDB', () => {
```
```suggestion
    describeIf(isMongoDB)('m:n mandatory (explicit) - MongoDB only', () => {
```

Jolg42 committed:
update README [skip ci]
